### DISCUSSION TITLE
Persistent mirai sessions

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#18] [Ideas]
Have you considered a persistent mirai where a single R session can send and receive multiple jobs? R startup time could be a bottleneck if there are numerous short-lived tasks. I am picturing something like `callr::r_session` as demonstrated at https://www.tidyverse.org/blog/2019/09/callr-task-q/. cc @krlmlr.

**Comment by shikokuchuo** [comment]
I have come across that article before. For my main use purposes R startup time is not really an issue as it is async anyway. But you probably have a specific use case in mind?

I am basically in the middle of re-writing the non-blocking call... a more optimal solution. So leave this with me for now...

**Comment by shikokuchuo** [comment]
So there is potentially an even lighter all-NNG solution to this. You see I don't really need a queue as NNG itself _is_ a queue in that all messages are buffered, and I believe there is a topology using the scalability protocols that achieves what you want. I just need to test it does what I think it does. 

Having to maintain so much state when working with NNG just felt intuitively wrong :)

I have limited bandwidth right now and re-implementing the non-blocking call is priority. But stay tuned...

**Comment by wlandau** [comment]
Great point, and that actually reminded me of how `clustermq` is actually almost a task queue: https://github.com/mschubert/clustermq. To be a task queue like Gabor's, I think it just needs (1) a way to keep a backlog of queued jobs, and (2) childproofing so users don't need to think about the ZeroMQ ping-pong event loop.

**Comment by shikokuchuo** [comment]
Re-writing the non-blocking call took much less time than expected. This is now handled entirely in C and as robust as any other part of the package.

Fortunately for you, there was also an internal request for similar functionality on our side so I have put together a prototype in 0d983b0393d7116e52460faa7dbccdc24600493e that might cover your bases as well.  NNG worked exactly as expected, hence implementation was straightforward.

Basic usage docs are included in the Rd file, but I thought it worth pointing out a few things:

Firstly, this offers a minimal approach.

If you were after a maximal approach where you want to monitor the status of each task, that _can_ also be achieved using NNG, and there is even an entire statistics interface _if_ you wanted to go there, but I think it is also against the whole ethos of NNG. The library was designed to handle concurrency precisely so we don’t need to ourselves.

Also as you mentioned ZeroMQ, I just wanted to make sure that you are aware that NNG is not ZeroMQ. There has been an evolution from ZeroMQ to Nanomsg to NNG, which is a complete re-write by Garrett D'Amore as it should have been in his view (https://nng.nanomsg.org/RATIONALE.html). The two are not compatible over the wire.

Your 2 points: backlog – yes. And optimal load-balancing / automatic distribution of tasks handled by NNG at the C library level. Processes can be created/destroyed at will. Child-proof –  no changes to existing functions. New `daemons()` function with minimal user API. 

You should be able to do something like:
```
daemons(4L)

x <- 8L
res <- vector(mode = "list", length = x)
start <- .Internal(Sys.time())
for (i in seq_len(x)) {
  res[[i]] <- mirai({
    st <- .Internal(Sys.time())
    .Internal(Sys.sleep(x))
    .Internal(Sys.time()) - st
  }, x = i)
}
.Internal(Sys.time()) - start
lapply(res, call_mirai)
.Internal(Sys.time()) - start
unlist(lapply(res, .subset2, "data"))

**Comment by shikokuchuo** [comment]
Just adding a note that the nanonext Aio user interface has been streamlined now the non-blocking call code has been finalised. 

This means a 'mirai' no longer needs to be expressly called. Just access the field `$data` directly (unfortunately this will be a breaking change from the current $value). It will return the evaluated result or an NA 'unresolved value'.

**Comment by shikokuchuo** [comment]
Got round to updating the package today after implementing a whole host of new features in nanonext. The new daemons 'control panel' now has a logical interface which is modular by design and I can live with.

Coming back to this request - the current prototype performs well for similar homogenous tasks, but is not a true queue. This is as NNG does load-balancing at the messaging level, and does not maintain the state of each request (whether each has received a reply or not). Coupled with buffering that is tricky to turn off and it is no longer trivial at this level.

Demonstrated below:

```
daemons(4L)

x <- c(1L, 2L, 6L, 5L, 8L, 7L, 3L, 4L)
res <- vector(mode = "list", length = length(x))
start <- .Internal(Sys.time())
for (i in seq_along(x)) {
  res[[i]] <- mirai({
    st <- .Internal(Sys.time())
    .Internal(Sys.sleep(x))
    .Internal(Sys.time()) - st
  }, x = x[i])
}
.Internal(Sys.time()) - start
lapply(res, call_mirai)
.Internal(Sys.time()) - start
unlist(lapply(res, .subset2, "data"))
```
The above should give you an optimal 9s, but the below a non-optimal 12s.

```
daemons(4L)

x <- c(10L, 5L, 1L, 1L, 1L, 1L, 1L, 1L, 1L)
res <- vector(mode = "list", length = length(x))
start <- .Internal(Sys.time())
for (i in seq_along(x)) {
  if(i == 5L) .Internal(Sys.sleep(6))
  if(i == 9L) .Internal(Sys.sleep(2))
  res[[i]] <- mirai({
    st <- .Internal(Sys.time())
    .Internal(Sys.sleep(x))
    .Internal(Sys.time()) - st
  }, x = x[i])
}
.Internal(Sys.time()) - start
lapply(res, call_mirai)
.Internal(Sys.time()) - start
unlist(lapply(res, .subset2, "data"))
```
A true 'queue' would allocate the later requests to the already finished and free processes.

**Comment by shikokuchuo** [comment]
It seems it wouldn't be too much work to add on a queue similar to Gabor's, but I wonder if that's really the right solution.

I think I would prefer a more decentralised design. The scheduler may broadcast when it has a new job and idle processes can respond perhaps (in a surveyor/respondent pattern in NNG terms). Even better if the scheduler sits in another thread in the background (NNG creates portable threads and I have tested passing SEXPs to them but I have not tried messaging from them!). This is just one of many possibilities.

But it seems the less state we track the more resilient the solution becomes. If a process dies but we have a dataframe that says it is idle...

**Comment by shikokuchuo** [comment]
@wlandau A simple approach as per the article seems to breaks down as it would only update upon user interaction. If for example you set 100 tasks to run and then go way you may come back to find that most of them remain queued.

The more general solution seems to necessitate C code to have the scheduler run on another thread. This would store the state of daemons along with pointers to the language objects that are awaiting execution. These would also have to be duplicated. Not sure if this is what you had in mind.

In any case I'm not going to hold up the next release of 'mirai' for this as the current setup will be useful in many cases and will likely remain one of the options going forward anyway.

**Comment by wlandau** [comment]
> A simple approach as per the article seems to breaks down as it would only update upon user interaction. If for example you set 100 tasks to run and then go way you may come back to find that most of them remain queued.

For a task queue like Gabor's, I would expect to need a custom event loop that polls the scheduler so queued tasks regularly ship to workers. Not prohibitive, but I agree, not ideal either.

> The more general solution seems to necessitate C code to have the scheduler run on another thread. This would store the state of daemons along with pointers to the language objects that are awaiting execution. These would also have to be duplicated. Not sure if this is what you had in mind.

Marshaling those jobs to a separate process with a separate event loop is tricky enough that I have not attempted it myself. Pretty convenient if it happens to exist already.

> In any case I'm not going to hold up the next release of 'mirai' for this as the current setup will be useful in many cases and will likely remain one of the options going forward anyway.

Glad to hear I am not slowing you down. Thanks for the discussion and features!

**Comment by shikokuchuo** [comment]
@wlandau You're welcome! mirai 0.1.1 is out. Turned out not to be 0.2.0 as no breaking changes.

Unfortunately the full task manager doesn't exist yet. My thoughts keep changing on this so I think I'll let this settle for a while. Contributing something to upstream might be the answer. I think all we need is basically messages to be buffered at the sender if the daemons are yet to reply. The need for this kind of pattern should be quite common.

**Comment by shikokuchuo** [comment]
@wlandau I have come back to this and the task queue is now implemented using a simple all-R solution. A separate process polls for changes every 5ms, perhaps a bit excessive... It is still a bit rough around the edges, so welcome any comments.

In any case, my previous example now returns the optimal result of (just over) 10s:

```
library(mirai)
daemons(4, q = TRUE)

x <- c(10L, 5L, 1L, 1L, 1L, 1L, 1L, 1L, 1L)
res <- vector(mode = "list", length = length(x))
start <- .Internal(Sys.time())
for (i in seq_along(x)) {
  if(i == 5L) .Internal(Sys.sleep(6))
  if(i == 9L) .Internal(Sys.sleep(2))
  res[[i]] <- mirai({
    st <- .Internal(Sys.time())
    .Internal(Sys.sleep(x))
    .Internal(Sys.time()) - st
  }, x = x[i])
}
.Internal(Sys.time()) - start
lapply(res, call_mirai)
.Internal(Sys.time()) - start
unlist(lapply(res, .subset2, "data"))
```

**Comment by wlandau** [comment]
This is fantastic, @shikokuchuo! Very powerful as a local multi-process task queue. 

Since we last spoke, I have been thinking more about the future direction of https://github.com/wlandau/crew. My goal is to build on top of an existing local-network task queue and make it easier to launch daemons/server processes on different kinds of platforms, from SLURM to AWS. From the update you shared, `mirai` seems much closer to being an alternative to https://github.com/mrc-ide/rrq which could support https://github.com/wlandau/crew/issues/21. It would be really nice to drop the dependency on Redis.

For `crew`, I would like to be able to auto-scale server processes in response to an increasing backlog of tasks. It would be great to be able to launch a server process on a remote node (e.g. as a SLURM job) which connects back to the client on the local network (`mirai::server()` can already do this, right?) Then if the workload subsides, it would be nice to automatically drop the workers that idle for too long. Seems like this would need some way of detecting which TCP sockets are open and working.

Would this set of features be possible or in scope for `mirai`? I know you intend `mirai` to be lightweight, so I totally understand if some or all of this will not be on your roadmap.

**Comment by shikokuchuo** [comment]
For the remote queue I think I just need to add another interface to specify an array of URLs for the task queue to listen to. Then each remote daemon will dial into its unique URL using the same `server()` function as before.

Autoscaling requires a bit more work: killing workers when they idle for too long is easy. But there is currently not the functionality to start up remote workers automatically within the package, as that would be use-dependent e.g. Slurm / AWS and I did not want to get into specifics when there are probably better existing tools.

It sounds like `crew` is designed to abstract over remote implementations so I think logically that functionality should sit there. I will have to think about how easy it will be to have a scalable queue size though as it is fixed at the moment.

**Reply by wlandau** [reply]
> Autoscaling requires a bit more work: killing workers when they idle for too long is easy. But there is currently not the functionality to start up remote workers automatically within the package, as that would be use-dependent e.g. Slurm / AWS and I did not want to get into specifics when there are probably better existing tools.
> 
> It sounds like crew is designed to abstract over remote implementations so I think logically that functionality should sit there. I will have to think about how easy it will be to have a scalable queue size though as it is fixed at the moment.

Exactly, I envision `crew` to focus on launching workers on specific platforms. This requires a way to find out the length of the queue, how many workers are alive, and how many workers can accept new tasks.

**Comment by shikokuchuo** [comment]
I just implemented (and reverted) the local queue / remote cluster setup as I think the interface becomes too unwieldy.

For now I'd like to keep the options as either all local, started automatically via `daemons(q = TRUE)`, or all remote (remote queue + remote cluster) simply by swapping out `server()` for `serverq()`.

The benefit from the current simplicity is that you can actually do autoscaling to some extent by adding `serverq()` clusters that dial into the same client URL and tasks are distributed evenly between them. This isn't perfect load-balancing of course but probably good enough in many cases and it does allow resources to be added or removed very easily.

**Reply by wlandau** [reply]
Do you have an examples using `server()` and `serverq()`? I'm not totally clear on how to use them, but from looking at the code I get the sense that they would be extremely useful and I might be able to take on the auto-scaling responsibilities in `crew`.

**Reply by shikokuchuo** [reply]
Is the readme example sufficient? https://shikokuchuo.net/mirai/#distributed-computing
If there’s a better way to frame some of this do let me know.

`server()` in a background process evaluates all mirai requests. It receives a message, evaluates it and then sends back the result. That is, as you can see from the few lines of code, all it does. 

`serverq()` essentially keeps track of 2 lists, a list of server processes, and a list of tasks and assigns tasks to free processes. It sits between the client and individual `server()` processes.

Both functions only require the client URL (to be able to pass messages). `serverq()` also maintains ‘n’ server connections each with a unique URL, but I have kept that out of the user interface for simplicity.

Hope this is useful. If there's an interface that's not currently provided for but will be useful for you do reach out.

**Reply by wlandau** [reply]
Thanks, that helps! So each server runs:

```r
server("tcp://LOCAL_IP_OF_CLIENT:PORT")
```

and the client runs:

```r
daemons(.url = "tcp://LOCAL_IP_OF_CLIENT:PORT")
m <- mirai(...)
m2 <- mirai(...)
```

I am beginning to see how the pieces would fit together, and I can definitely take responsibility in `crew` to make sure `server()` runs on e.g. SLURM or AWS. The possibilities are really exciting, and I have asked my devops team to install `nanonext` and NNG on my company's cluster so I can test this. I do have follow-up questions:

1. Above, is `LOCAL_IP_OF_CLIENT` the same in all calls to `server()` and `daemons()`? Can I get that IP by calling `ifconfig` on the client node? 
2. Is `q = TRUE` in `daemons()` compatible with `server()` above?
3. How exactly do I use `serverq()`? Is it a replacement for `server()` if I use `q = TRUE` in `daemons()`?
4. Is there a way to detect crashed servers? At minimum for my use case, the client would need to know immediately if a server quit unexpectedly. (Heartbeating?) At best (but not required) the client would be able to identify a task that failed so `crew` could retry it.
5. How does `daemons_view()` work? Can it poll the URL to tell how many servers are running and have active working connections to the TCP socket? I would definitely need some way to know how many servers are running, and I would trust that information more if it came directly from NNG TCP sockets (as opposed to internal bookkeeping that `crew` might do.)
6. If 1-5 are all doable, would you be open to implementing functionality in `server()` (or would I use `serverq()` here?) to automatically scale down a worker if the workload decreases? It would be super nice to have:
    A. An optional timeout so a worker quits if it idles too long. (Maybe with a grace period when it initializes so it can accept its first task?)
    B. An optional upper limit for the number of tasks `server()` before it exits.

**Reply by shikokuchuo** [reply]
>  I do have follow-up questions:
> 
>     1. Above, is `LOCAL_IP_OF_CLIENT` the same in all calls to `server()` and `daemons()`? Can I get that IP by calling `ifconfig` on the client node?
Yes to both. You can use an IPv4 or v6 address.
>     2. Is `q = TRUE` in `daemons()` compatible with `server()` above?
q = TRUE in `daemons()` only applies to the local case. If specifying ‘.url’ for use with a remote server then this is not applicable.
>     3. How exactly do I use `serverq()`? Is it a replacement for `server()` if I use `q = TRUE` in `daemons()`?
Just use as a replacement for `server()`, specify ‘.url’ in `daemons()` rather than 'q'.
>     4. Is there a way to detect crashed servers? At minimum for my use case, the client would need to know immediately if a server quit unexpectedly. (Heartbeating?) At best (but not required) the client would be able to identify a task that failed so `crew` could retry it.
 - processes may crash for whatever reason, but if just down to code failure, all execution is wrapped in a tryCatch, so all errors (and interrupts) are returned as such to the mirai, and the process just continues. See https://shikokuchuo.net/mirai/#errors-interrupts-and-timeouts
 - the network scales automatically. So if a server does die, or new servers connect to the client URL, tasks are distributed automatically.
 - Heartbeating is horrible. I would guess it would be ok to check every time a new request is sent? See the answer below for a solution that doesn’t require polling.
>     5. How does `daemons_view()` work? Can it poll the URL to tell how many servers are running and have active working connections to the TCP socket? I would definitely need some way to know how many servers are running, and I would trust that information more if it came directly from NNG TCP sockets (as opposed to internal bookkeeping that `crew` might do.)
-  daemons_view()` just queries the internal account of the number of active sockets (what you specify as ‘n’. This should be accurate when running locally, but of course if specifying you are gong to have 4 servers connecting at ‘.url’, you actually have to start those 4 server instances yourself and there is no guarantee there.
- There are two potential ways to do this: best is querying stats recorded at the NNG level (I would need to look into this to check it exists and then build the interface in `nanonext`). An alternative would involve creating another socket in each process using a surveyor/respondent protocol where the client sends a timed survey and the servers respond. However we will have to launch this on another thread within the process such that this channel is always available and not blocked by the R thread. I would also need to implement this in `nanonext`, and it adds overhead to each process.
>     6. If 1-5 are all doable, would you be open to implementing functionality in `server()` (or would I use `serverq()` here?) to automatically scale down a worker if the workload decreases? It would be super nice to have:
>        A. An optional timeout so a worker quits if it idles too long. (Maybe with a grace period when it initializes so it can accept its first task?)
>        B. An optional upper limit for the number of tasks `server()` before it exits.
- Adding an interface in `server()` is possible for an optional timeout so if it waits for too long it exits.
- Similarly, it is possible to add something in `serverq()` that kills off individual servers if they idle for too long.

I can also see that the current naming is causing some confusion and I will have to think of something better than `serverq` and `q = TRUE`...

**Reply by shikokuchuo** [reply]
perhaps if I replace `serverq()` with `cluster()`, so you start either a server or a cluster. More intuitive?

**Reply by wlandau** [reply]
> daemons_view()` just queries the internal account of the number of active sockets (what you specify as ‘n’. This should be accurate when running locally, but of course if specifying you are gong to have 4 servers connecting at ‘.url’, you actually have to start those 4 server instances yourself and there is no guarantee there.

For my auto-scaling use case specifically, I do not think I will use `n` because I will be launching the workers outside `mirai` and the number of workers will change due to changing task load or crashes.

> Heartbeating is horrible. I would guess it would be ok to check every time a new request is sent? See the answer below for a solution that doesn’t require polling... best is querying stats recorded at the NNG level (I would need to look into this to check it exists and then build the interface in nanonext).

 I do wonder how NNG handles this. If NNG does have a way to reliably detect active connections / crashes, it would be amazing to have access to this in `mirai`. 

> perhaps if I replace serverq() with cluster(), so you start either a server or a cluster. More intuitive?

A new name might help, I am not sure. I think it would be more helpful to read a description from first principles and a runnable example. I am still having trouble understanding what it does and how it is different from `server()`.

Thanks for all your help answering my questions.

**Reply by shikokuchuo** [reply]
Great, I have gone ahead and implemented the stats interface in nanonext. Now `daemons_view()` will report an accurate number of connections.

I've reverted the daemons interface as I think it was becoming confusing when to use 'n' together with '.url'. Now I hope it's clearer that it's one or the other. I have also changed the behaviour slightly and documented as such to make it more consistent.

I'll leave it as is for you to explore. If you think of any changes to make it more intuitive do let me know.

You'll need dev versions of both nanonext and mirai, which you can get from r-universe:
```r
options(repos=c(shikokuchuo='https://shikokuchuo.r-universe.dev',CRAN ='https://cloud.r-project.org')
install.packages(c("nanonext", "mirai"))
```

Sorry  I couldn't be very expressive on the examples this is mainly to prevent possible problems during CRAN tests - it is currently listed as --no-tests as it is... I will try to add to the README when I get around to it, but everything relevant should be documented in each function itself.

**Reply by shikokuchuo** [reply]
> I do wonder how NNG handles this. If NNG does have a way to reliably detect active connections / crashes, it would be amazing to have access to this in `mirai`.

Made me think of  this quote from Simon Urbanek on R-devel mailing list from a couple of days ago:

> Sockets and I/O are quite complex matter - it's easy to get it wrong and create hard-to-detect bugs in you code unless you are an expert in it. It's one of the wheels you don't want to be reinventing.

**Reply by wlandau** [reply]
Thank you so much for these updates, @shikokuchuo. I am eager to continue experimenting. (Apologies in advance if this next phase takes me a while, I may be pulled aside for other work in the next couple weeks.)

> Sorry I couldn't be very expressive on the examples this is mainly to prevent possible problems during CRAN tests - it is currently listed as --no-tests as it is... I will try to add to the README when I get around to it, but everything relevant should be documented in each function itself.

I understand. I have received feedback in the past about excessive use of `\dontrun{}`, and I am not sure about `\donttest{}`. If it helps, most of the examples in `targets` are inside an ` if (identical(Sys.getenv("TAR_EXAMPLES"), "true")) {}` block, which makes them available to users but skipped on CRAN. I have never received negative feedback from CRAN about this approach.

**Reply by shikokuchuo** [reply]
That's good to know. I had included something like that for the tests as some would have violated the CRAN 2 thread limit. My fear was that they find ways of testing the examples anyway in their ever expanding special case test machines... have you been bitten by black swan testing? UBSAN? :)

Anyhow I have made some minor changes around the edges - in particular folded `daemons_view()` back into `daemons()` - only one function to remember! All clearly documented I hope.

**Reply by shikokuchuo** [reply]
Similarly, I've folded `serverq()` into `server()` with an argument 'n'.

**Reply by wlandau** [reply]
> have you been bitten by black swan testing? UBSAN? :)

Not yet, fortunately.

**Reply by wlandau** [reply]
I am experimenting with `mirai` on an SGE cluster, and it's going super well! On the client node, I ran

```r
library(mirai)
daemons("tcp://CLIENT_IP:5555")
while(daemons()$connections < 2) {
  Sys.sleep(0.1)
}
m1 <- mirai({
  Sys.sleep(5)
  "worker1"
})
m2 <- mirai({
  Sys.sleep(5)
  "worker2"
})
for (i in 1:20) {
  Sys.sleep(0.5)
  print(sprintf("%ss %s %s", i / 2, m1$data, m2$data))
}
```

and on each of two different server nodes:

```r
library(mirai)
server("tcp://CLIENT_IP:5555")
```

After each server process initializes and `daemons()$connections` returns 2, the client exits the Sys.sleep(0.1) loop and I see the following:

```r
[1] "0.5s NA NA"
[1] "1s NA NA"
[1] "1.5s NA NA"
[1] "2s NA NA"
[1] "2.5s NA NA"
[1] "3s NA NA"
[1] "3.5s NA NA"
[1] "4s NA NA"
[1] "4.5s NA NA"
[1] "5s worker1 worker2"
```

So it looks like the 2 tasks are distributed in parallel without needing to use the `n` argument of `server()` (which I still don't understand, but I guess I don't need for my use case?)

Anyway, this is nearly perfect! All I need is automatic down-scaling in `server()` and a way to detect crashes/lost tasks. I will post separate issues for those.

**Reply by shikokuchuo** [reply]
That's great! I see you picked up the latest version. Sorry for working in parallel but the interface I'm now happy with.

**Reply by shikokuchuo** [reply]
You need the active queue only when your tasks exceed the number of workers. Simple example: 2 workers, 3 jobs: 10s, 5s, 5s.

If  you connect 2 `server()` instances your client distributes the 10s job to worker 1, 5s job to worker 2, and the last 5s job to worker 1 again. So in total it would take 15s.

If you connect one `server(n = 2)`,  the active queue will allocate the last 5s job to worker 2 as soon as it's finished its first job, so your total time will now only be 10s.

**Reply by shikokuchuo** [reply]
If the length of your jobs are similar then you also wouldn't care, and would go for the simpler implementation.

**Reply by wlandau** [reply]
Ah, so it's a load balancing issue. That is important for my situation. I will have many more tasks than workers, and the tasks will vary in execution time. 

In my situation, each server will run on its own compute node on the local network. In the diagram below, I would need the load balancing to happen between Server 1 and Server 2.

![Screenshot 2023-02-22 at 3 44 08 PM](https://user-images.githubusercontent.com/1580860/220754507-ebc2bc81-d745-4201-9328-748d3b3edc10.png)

If I use the `n` argument, the diagram would look like this:

![Screenshot 2023-02-22 at 3 48 06 PM](https://user-images.githubusercontent.com/1580860/220755308-caf66740-387b-4ac9-b4d5-eb99aba0d107.png)

If I understand correctly, that will launch 2 parallel daemons inside each server, and then the load balancing only happens among the multicore processes within each respective server. I don't think that would fit the auto-scaled distributed and cloud computing case I am going for.

Can `daemons(“tcp://192.168.0.2:5555”, q)` perform load balancing across servers when there are more (and variable) tasks than servers?

![Screenshot 2023-02-22 at 3 52 39 PM](https://user-images.githubusercontent.com/1580860/220756317-7de96eac-7484-4cd7-8db1-33b6d3be2f84.png)

**Reply by shikokuchuo** [reply]
Yes, you're in luck. That is possible. Nice idea with the `daemons("IP_ADDR", q)` interface. Let me have a go at writing it!

**Reply by wlandau** [reply]
Fantastic! I look forward to trying it out when you are ready.

**Reply by shikokuchuo** [reply]
Will let you know. I've taken the important first step of allowing `daemons()` to pass arguments through to the `server()` function. Have also introduced the notion of 'nodes' to make it conceptually clearer (I hope).

**Reply by wlandau** [reply]
Thank you so much for working on this! 

From the documentation in your latest commit, I see

> #' @param nodes [default NULL] if supplied, this server instance will run as an
 #'     active queue (task scheduler) with the specified number of nodes.

Would you elaborate? I think of each `server()` as its own node, as opposed to having multiple nodes within the server. For the load balancing and active queue, we are still talking about delegate among `server()`s and not among processes within a given `server()`, right? Does a server need to know the number of nodes there are?

> I've taken the important first step of allowing daemons() to pass arguments through to the server() function.

Thanks! Which arguments will be passed? For my case, sometimes `server()` will be called on a remote computer on the cluster long after `daemons()` was called on the client IP.

**Reply by shikokuchuo** [reply]
This will all become apparent as the interface is implemented. Just to note that if you run an active queue, no matter where the nodes are, the queue will be it's own process, not the actual user session.

The end result is to have the client connect via the local queue to remote nodes.

**Reply by wlandau** [reply]
I am testing the load balancing on an SGE cluster (see below), and so far so good! But I am still unclear what is happening internally. If the client runs `daemons("tcp://xx.xx.xx.138:5555", nodes = 2)`, and then two different computers on the local network each run `server("tcp://xx.xx.xx.138:5555")`, then which server actually runs the active queue? (Do they both?)

Also, just a note that I think I need to call `daemons()` before I call `server()` in order for the active queue to work. I can definitely work with that, but I would not have known without testing it. If I call `daemons()` first:

```{r}
# On xx.xx.xx.138:
library(mirai)
daemons("tcp://xx.xx.xx.138:5555", nodes = 2)

# On xx.xx.xx.175 and xx.xx.xx.176:
library(mirai)
server("tcp://xx.xx.xx.138:5555")

# Again, on xx.xx.xx.138:
m1 <- mirai({
  Sys.sleep(5)
  paste0("job1_", getip::getip())
})
m2 <- mirai({
  Sys.sleep(5)
  paste0("job2_", getip::getip())
})
m3 <- mirai({
  Sys.sleep(5)
  paste0("job3_", getip::getip())
})
m4 <- mirai({
  Sys.sleep(5)
  paste0("job4_", getip::getip())
})
for (i in 1:50) {
  Sys.sleep(1)
  print(sprintf("%ss %s %s %s %s", i, m1$data, m2$data, m3$data, m4$data))
}

# Output on xx.xx.xx.138:
[1] "1s NA NA NA NA"
[1] "2s NA NA NA NA"
[1] "3s NA NA NA NA"
[1] "4s NA NA NA NA"
[1] "5s job1_xx.xx.xx.175 job2_xx.xx.xx.176 NA NA"
[1] "6s job1_xx.xx.xx.175 job2_xx.xx.xx.176 NA NA"
[1] "7s job1_xx.xx.xx.175 job2_xx.xx.xx.176 NA NA"
[1] "8s job1_xx.xx.xx.175 job2_xx.xx.xx.176 NA NA"
[1] "9s job1_xx.xx.xx.175 job2_xx.xx.xx.176 NA NA"
[1] "10s job1_xx.xx.xx.175 job2_xx.xx.xx.176 job3_xx.xx.xx.175 job4_xx.xx.xx.176"
```

If I call `server()` first:

```{r}
# On xx.xx.xx.175 and xx.xx.xx.176:
library(mirai)
server("tcp://xx.xx.xx.138:5555")

# Then on xx.xx.xx.138:
library(mirai)
daemons("tcp://xx.xx.xx.138:5555", nodes = 2)
m1 <- mirai({
  Sys.sleep(5)
  paste0("job1_", getip::getip())
})
m2 <- mirai({
  Sys.sleep(5)
  paste0("job2_", getip::getip())
})
m3 <- mirai({
  Sys.sleep(5)
  paste0("job3_", getip::getip())
})
m4 <- mirai({
  Sys.sleep(5)
  paste0("job4_", getip::getip())
})
for (i in 1:50) {
  Sys.sleep(1)
  print(sprintf("%ss %s %s %s %s", i, m1$data, m2$data, m3$data, m4$data))
}

# Output on xx.xx.xx.138:
[1] "1s NA NA NA NA"
[1] "2s NA NA NA NA"
[1] "3s NA NA NA NA"
[1] "4s NA NA NA NA"
[1] "5s job1_40.1.29.176 NA NA NA"
[1] "6s job1_40.1.29.176 NA NA NA"
[1] "7s job1_40.1.29.176 NA NA NA"
[1] "8s job1_40.1.29.176 NA NA NA"
[1] "9s job1_40.1.29.176 NA NA NA"
[1] "10s job1_40.1.29.176 job2_40.1.29.176 NA NA"
[1] "11s job1_40.1.29.176 job2_40.1.29.176 NA NA"
[1] "12s job1_40.1.29.176 job2_40.1.29.176 NA NA"
[1] "13s job1_40.1.29.176 job2_40.1.29.176 NA NA"
[1] "14s job1_40.1.29.176 job2_40.1.29.176 NA NA"
[1] "15s job1_40.1.29.176 job2_40.1.29.176 job3_40.1.29.176 NA"
[1] "16s job1_40.1.29.176 job2_40.1.29.176 job3_40.1.29.176 NA"
[1] "17s job1_40.1.29.176 job2_40.1.29.176 job3_40.1.29.176 NA"
[1] "18s job1_40.1.29.176 job2_40.1.29.176 job3_40.1.29.176 NA"
[1] "19s job1_40.1.29.176 job2_40.1.29.176 job3_40.1.29.176 NA"
[1] "20s job1_40.1.29.176 job2_40.1.29.176 job3_40.1.29.176 job4_40.1.29.176"
```

**Reply by wlandau** [reply]
I also tested the case when there are 3 servers and 2 "nodes", but the load balancing still happens across the 3 servers. It's a positive outcome because the servers will scale up and down in a `targets` pipeline, but I am confused about what `nodes = 2` really means. There is something I am missing about how the active queue works.

```r
# On the client process in xx.xx.xx.138:
library(mirai)
daemons("tcp://xx.xx.xx.138:5555", nodes = 2)

# On xx.xx.xx.175, xx.xx.xx.176, and a separate R process on xx.xx.xx.138:
library(mirai)
server("tcp://xx.xx.xx.138:5555")

# Again, in the client process on xx.xx.xx.138:
daemons("tcp://40.1.28.138:5555", nodes = 2) # note nodes = 2 even though there are 3 servers
m1 <- mirai({
  Sys.sleep(5)
  paste0("job1_", getip::getip())
})
m2 <- mirai({
  Sys.sleep(5)
  paste0("job2_", getip::getip())
})
m3 <- mirai({
  Sys.sleep(5)
  paste0("job3_", getip::getip())
})
m4 <- mirai({
  Sys.sleep(5)
  paste0("job4_", getip::getip())
})
for (i in 1:50) {
  Sys.sleep(1)
  print(sprintf("%ss %s %s %s %s", i, m1$data, m2$data, m3$data, m4$data))
}

# Output in the client process on xx.xx.xx.138: 3 jobs run concurrently on the 3 servers:
[1] "1s NA NA NA NA"
[1] "2s NA NA NA NA"
[1] "3s NA NA NA NA"
[1] "4s NA NA NA NA"
[1] "5s job1_xx.xx.xx.138 job2_xx.xx.xx.175 job3_xx.xx.xx.176 NA"
[1] "6s job1_xx.xx.xx.138 job2_xx.xx.xx.175 job3_xx.xx.xx.176 NA"
[1] "7s job1_xx.xx.xx.138 job2_xx.xx.xx.175 job3_xx.xx.xx.176 NA"
[1] "8s job1_xx.xx.xx.138 job2_xx.xx.xx.175 job3_xx.xx.xx.176 NA"
[1] "9s job1_xx.xx.xx.138 job2_xx.xx.xx.175 job3_xx.xx.xx.176 NA"
[1] "10s job1_xx.xx.xx.138 job2_xx.xx.xx.175 job3_xx.xx.xx.176 job4_xx.xx.xx.138"
```

**Reply by shikokuchuo** [reply]
I'll get around to looking at your examples, but the local active queue / remote nodes is implemented in mirai 0.7.2.9005. You see where I was going? :) Docs and readme are updated. It's likely I am out until next week so I'll leave it with you for now.

**Reply by wlandau** [reply]
Thanks again for your work on all this. Please enjoy your time away.

**Reply by wlandau** [reply]
Thanks for the updates. I read through https://github.com/shikokuchuo/mirai#example-2-connecting-to-remote-servers-through-a-local-server-queue, and while I like the idea of a local active queue, I hesitate to use it in my situation because of the large number of ports it would create.

My use case is an excellent fit with the interface already https://github.com/shikokuchuo/mirai#example-1-connecting-to-remote-servers--remote-server-queues. I still do not understand the parts about nodes and active queues, but I will start a separate thread about that.

**Reply by shikokuchuo** [reply]
Parking your last comment, which I will answer in the new thread you set up.

In reply to your previous questions - I have now tweaked the interface so it works how you wanted it to in your examples. This is v0.7.2.9006. The active queues should now run as intended. README and docs are updated.

To really see the effect of the active queue - set the sleep time of the first job to 15s.

There is no difference between starting `daemons()` first or the servers first. On a lower level, the servers dial in to the client. By default this is async and will continue retrying if there is no client listening at the URL. I have enabled a new argument `asyncdial` which can be useful for troubleshooting connection issues. If set to `FALSE`, it will error if a connection cannot be established immediately - e.g. port not open, or `daemons()` has not been called yet.

Please take note I have tweaked some of the other arguments as well. You can also adjust the polling rate of the active queue should you need to - I am not sure what sort of defaults you had in mind.

**Comment by shikokuchuo** [comment]
The end result of this year-long dialogue: https://cran.r-project.org/package=crew
